broadinstitute/single_cell_portal_core
# Allowing for multiple collection curators, studies in multiple collections (SCP-3849, SCP-3867, SCP-3872)

Currently, each collection may only have one "curator" (user who controls assigning studies to collections), and a study may only belong to one collection.  Also, curators are not allowed to edit the collection themselves, and must go through an SCP admin to make any updates to the collection itself (like the tag line, or visual characteristics).

This update allows for collections to have multiple curators (who can now directly edit the collection), and for studies to belong to multiple collections.  This is done by changing the associations from `has_many -> belongs_to` to `has_and_belongs_to_many` for the `Study`, `BrandingGroup` (collections), and `User` models.  This way, we can access all the studies for a collection directly from the collection itself, and vice versa.  This update also includes a new integration test suite to cover all new functionality.  In addition, a reversible data migration has been added to migrate from the existing association structure to the new one.

There is also now a visual indication on the Collections page as to whether a collection has been published or not.  This only shows if a user is signed in, and is a curator for at least one collection (otherwise, the only collections visible will all be published, and the indicator is not necessary).  Example:

![Screen Shot 2021-11-18 at 10 30 39 AM](https://user-images.githubusercontent.com/729968/142445508-98de60e3-30b3-40e1-a310-19d1bb55437a.png)

MANUAL TESTING
1. Pull this branch, and run the migration via `rails db:migrate`
2. Enter the Rails console and run `SyntheticBrandingGroupPopulator.populate_all`
3. Sign in as an admin, and select "Branding Groups" from the profile menu
4. For each new synthetic collection, in the "curators" text field, add another non-admin user account:
![Screen Shot 2021-11-16 at 6 11 25 PM](https://user-images.githubusercontent.com/729968/142080914-e358c3be-d493-4a06-817e-bc22a0d3609d.png) 
5. Sign out, and sign in as the non-admin
6. Go to the "My Studies" page, and add any study to both new collections at the same time
7. From the "Browse Collections" page, select both new collections and ensure that they contain the study you just added
8. From the "Browse Collections" page, confirm that the non-admin account can see & click the "Edit" button for the new collections:
![Screen Shot 2021-11-18 at 10 28 48 AM](https://user-images.githubusercontent.com/729968/142445201-04f73c04-4494-4674-bf5b-ce53c3d897fc.png)
 
This PR satisfies SCP-3849, SCP-3867, and SCP-3872.
